lane.py

```diff
@@ -2138,99 +1664,325 @@ def _featured_works_with_lanes(
+
+        if qu._distinct is False:
+            # This query must always be made distinct, since a Work
+            # can have more than one LicensePool. If no one else has
+            # taken the opportunity to make it distinct (e.g. the
+            # faceting object, while setting sort order), we'll make
+            # it distinct based on work ID.
+            qu = qu.distinct(Work.id)
+
+        # Allow the pagination object to modify the database query.
+        if pagination is not None:
+            qu = pagination.modify_database_query(_db, qu)
+
+        return qu
+
+    def base_query(self, _db):
+        """Return a query that contains the joins set up as necessary to
+        create OPDS feeds.
+        """
+        qu = _db.query(
+            Work
+        ).join(
+            Work.license_pools
+        ).join(
+            Work.presentation_edition
         )
 
-        # Allow some overage to reduce the risk that we'll have to
-        # use a given book more than once in the overall feed. But
-        # set an upper limit so that a weird random distribution
-        # doesn't retrieve far more items than we need.
-        lane_query = lane_query.limit(target_size*1.3)
-        return lane_query
-
-    def _restrict_query_to_window(self, query, target_size, facets):
-        """Restrict the given SQLAlchemy query so that it matches
-        approximately `target_size` items.
-
-        DEPRECATED This should not be necessary anymore.
-        """
-        if query is None:
-            return query
-        window_start, window_end = self.featured_window(target_size, facets)
-        if window_start > 0 and window_start < 1:
-            query = query.filter(
-                mw.random <= window_end,
-                mw.random >= window_start
+        # Apply optimizations.
+        qu = self._modify_loading(qu)
+        qu = self._defer_unused_fields(qu)
+        return qu
+
+    def bibliographic_filter_clauses(self, _db, qu):
+        """Create a SQLAlchemy filter that excludes books whose bibliographic
+        metadata doesn't match what we're looking for.
+
+        :return: A 2-tuple (query, clauses).
+
+        - query is either `qu`, or a new query that has been modified to
+        join against additional tables.
+        """
+        # Audience language, and genre restrictions are allowed on all
+        # WorkLists. (So are collection restrictions, but those are
+        # applied by only_show_ready_deliverable_works().
+        clauses = self.audience_filter_clauses(_db, qu)
+        if self.languages:
+            clauses.append(Edition.language.in_(self.languages))
+        if self.media:
+            clauses.append(Edition.medium.in_(self.media))
+        if self.fiction is not None:
+            clauses.append(Work.fiction==self.fiction)
+        if self.license_datasource_id:
+            clauses.append(
+                LicensePool.data_source_id==self.license_datasource_id
+            )
```

Comment:
The license_datasource_id code is new. The fiction and inherit_parent_restrictions code was merged in from this method's now-removed implementation in `Lane`.